### PR TITLE
[CONTINT-4437] Add Support for EndpointSlices

### PR DESCRIPTION
**Pull request by ewoodthomas**
### What does this PR do?

1. Adds support for Kubernetes EndpointSlices to the metadata controller.

2. Also fixes an uncaught bug in Endpoints implementation. The services are stored in a metaBundle specific to the node. If a service's addresses corresponding to a node were completely deleted without the service being deleted, we never remove the `kube_service` tag and continue to store the mapping of node -> pod -> tag in memory. This probably occurs infrequently, but it would lead to incorrect tagging and a potential memory leak.

### Motivation

https://datadoghq.atlassian.net/browse/CONTINT-4437

Customers reported losing the `kube_service` tag on the majority of their pods. Services create an Endpoints objects that maps the service to addresses on Pods. These Endpoints objects can hold a maximum of 1000 addresses. We use the Endpoints to tag pods with the kube_service tag, but if a service is tied to 1000+ pods the tag is lost.

https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/

As of k8s 1.19+, EndpointSlices are created whenever an Endpoints object is created. EndpointSlices store up to 100 addresses, but Kubernetes creates as many EndpointSlices as necessary. By supporting EndpointSlices, the `kube_service` tag will not be lost.

### Describe how you validated your changes
Added unit tests that cover all cases of adding, deleting, and updating endpointSlices. Also added some cases for Endpoints to catch previously uncaught errors.

QA
1. Start an EKS Auto-Mode cluster (pretty convenient for big deployments)
2. Deploy the agent with the following config options enabled
```
datadog:
  env:
    - name: DD_KUBERNETES_USE_ENDPOINT_SLICES
      value: true

agents:
  priorityClassCreate: true
  ...
```
3. Edit the cluster agent clusterrole `kubectl edit clusterrole datadog-agent-cluster-agent` and add
```
rules
  ...
  - apiGroups:
    - discovery.k8s.io
    resources:
    - endpointslices
    verbs:
    - get
    - list
    - watch
  ...
```
4. Deploy a 1000+ pod deployment tied to a single service
```
---
apiVersion: v1
kind: Namespace
metadata:
  name: test-namespace
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test-deployment
  namespace: test-namespace
  labels:
    app: test-app
spec:
  replicas: 1050
  selector:
    matchLabels:
      app: test-app
  template:
    metadata:
      labels:
        app: test-app
    spec:
      containers:
        - name: sleep-container
          image: alpine
          imagePullPolicy: IfNotPresent
          command: ["/bin/sh", "-c", "sleep infinity"]
          resources:
            requests:
              memory: "5Mi"
              cpu: "1m"
            limits:
              memory: "10Mi"
              cpu: "2m"
---
apiVersion: v1
kind: Service
metadata:
  name: test-service
  namespace: test-namespace
spec:
  selector:
    app: test-app
  ports:
    - protocol: TCP
      port: 80
      targetPort: 80
  clusterIP: None
```
5. Plot the metric `kubernetes.pods.running` and tag by `kube_service`
6. All 1050 pods should be tagged
7. Repeat with `DD_KUBERNETES_USE_ENDPOINT_SLICES = false`
8. Verified that changes did not impact endpoint implementation
9. Confirmed problem - exactly 1000 pods were tagged
10. Confirmed deletion worked for both

[Link to dashboard](https://dddev.datadoghq.com/dashboard/u2s-swx-tzf?fromUser=true&refresh_mode=paused&from_ts=1734018483846&to_ts=1734020283846&live=false) showing tests for both Endpoints and EndpointSlices implementations.

Screenshot 2024-12-12 at 5 18 49 PM

### Additional Notes

### CODE REVIEW

**Review by michaelcretzman — APPROVED** [review]

**Review by GustavoCaso — COMMENTED** [review]
LGTM for the ASC owned files

**Review by GustavoCaso — APPROVED** [review]

File: pkg/util/kubernetes/apiserver/controllers/metadata_controller_test.go [hunk 1]
```diff
@@ -310,6 +363,304 @@ func TestMetadataControllerSyncEndpoints(t *testing.T) {
...
+					"default": {
+						"pod2_name": sets.New("svc2"),
+					},
+				},
+			},
+			1,
+		},
+	}
+
+	for i, tt := range tests {
```
Comment (gabedos) [inline pkg/util/kubernetes/apiserver/controllers/metadata_controller_test.go:616]: nit: can we run the tests using the following pattern. will be more clear output for debugging when subtests fail.
```
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
                        ... <test_contents> ...
		})
	}
```

**Review by gabedos — APPROVED** [review]
LGTM. Synced in person and implementation + tests makes sense.

**Review by sblumenthal — APPROVED** [review]

**Review by aureleoules — APPROVED** [review]

**Pull request merged**